homeassistant/components/recorder/__init__.py

```diff
@@ -734,26 +735,29 @@ def _empty_queue(event):
 
         self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_FINAL_WRITE, _empty_queue)
 
-        def shutdown(event):
+        @callback
+        def _async_shutdown(event: Event) -> None:
             """Shut down the Recorder."""
-            if not hass_started.done():
-                hass_started.set_result(shutdown_task)
+            if not self._hass_started.done():
+                self._hass_started.set_result(SHUTDOWN_TASK)
             self.queue.put(StopTask())
-            self.hass.add_job(self._async_stop_queue_watcher_and_event_listener)
-            self.join()
+            self._async_stop_queue_watcher_and_event_listener()
+            self.hass.async_add_executor_job(self.join)
```

Comment:
Is there a reason we don't wait for this to finish?

Reply:
No good one 😄 .  It is an implementation error.

Reply:
I was really hoping that was the reason for the test failure, but still failing on the CI only and passing locally.